### ISSUE TITLE
networking: VPN support

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#16992]
It would be great to have a VPN section on the Networking page.

This should be like the Firewall functionality, where it's exposed on the Networking page yet has additional details on a sub-page.

It could include Wireguard and OpenVPN.

**Comment by garrett** [comment]
Yes, It'd be nice for Cockpit to be able to create and manage VPNs. 

You're right that a good first step is to at least show that they exist (they do exist as another network currently, as you demonstrate @ https://github.com/cockpit-project/cockpit/issues/16948#issuecomment-1039289225) and we should visually tag them somehow as such.

**Comment by gil-obradors** [comment]
Hi! I'm interested on make contributions from Newtork views. I'm currently working on NetDevOps, but recently met cockpit.
I keep an eye on that thread. In June I finish exams from university and I will have time for contributions!

**Comment by garrett** [comment]
@gil-obradors: That's great!

If you happen to know about IRC and how to use that with an IRC client, we're on #cockpit on irc.libera.chat.

Otherwise, there's a web-based way of joining and chatting with us using your browser at this link:
https://kiwiirc.com/nextclient/irc.libera.chat/#cockpit (Although you do have to become a registered user to actually join the channel. Libera has info about how to do that @ https://libera.chat/guides/registration)

When you do have the time (I suppose in June?), please drop by and we'd be happy to help.

We also have contribution guidelines and documentation on our website @ https://cockpit-project.org/external/wiki/Contributing.html, which also includes how to set up your system to work on Cockpit: https://cockpit-project.org/external/source/HACKING.html

**Comment by gil-obradors** [comment]
Thanks @garrett ! I will be full-free after 18 of June.
From now, in some idle times I'm going to study all the documentation,thanks for links. I'm amazed at that wonderful on demand architecture. I want to play on it! 
We're in touch!

**Comment by gil-obradors** [comment]
![work-done](https://user-images.githubusercontent.com/6016929/175863646-d31e7632-27d2-4aa8-9f03-b8c20997d81f.gif)

**Comment by gil-obradors** [comment]
Thanks @jelly for your time!

As do you know cockpit in deep, in OpenVPN scenario We need files with keys. Is there any solution better than specify the file path from the system? ( this implies having transfered the file before, Wireguard is more frendly in that... )

I'm on it! I have work quite advanced... maybe this weekend I can take it out of the oven :meat_on_bone: 

About networking vpn layout page... may be @garrett could you show de way... There are many possibilities:
- Dropdown with [openvpn,wireguard,pppoe] in actual interfaces table. ( Or better.... if vpn-provider not installed -> Install Wireguard, otherwise button with Wireguard text as attached image)
- Table order, order by type, different colors shadows , icons column
- New article networking-interfaces-vpn with new card-table

I don't care what we decide, I'm not very good at UI but I have time :smile: 

![imatge](https://user-images.githubusercontent.com/6016929/176237080-6f989b89-21f8-4bc3-a332-d979cdae835f.png)

**Comment by jelly** [comment]
> Thanks @jelly for your time!
> 
> As do you know cockpit in deep, in OpenVPN scenario We need files with keys. Is there any solution better than specify the file path of the system? ( this implies having transfered the file before, Wireguard is more frendly in that... )

Do you mean the remote or the local system?  We do have `FileAutoComplete` which is used in for example `cockpit-podman` to select a container path on the host.

**Comment by garrett** [comment]
I think that Add VPN button with a dropdown menu is reasonable. I'm not really a fan of that pattern, but it sure beats tossing a ton of various VPN buttons on the page.

We could just have add VPN with a selector too. However, the dialogs are pretty huge, especially the PPoE one:

![image](https://user-images.githubusercontent.com/10246/176485887-5a14fcf5-32ac-43ba-b8fc-e0d40cca798e.png)

- Are any of the options optional? What do they mean? 
- What's the difference between "Name" and "Service name"? I'm guessing name is the "Connection name". Leaving it unprefixed (just "Name") is a bit confusing, especially with "Username" right below it. (It could be _your_ name if it's just an empty field.)
- What type of values are needed for each? (I'm assuming the LCP stuff is numbers...)
- Shouldn't the parent interface be a dropdown with available (and probably active) interfaces?

**Comment by gil-obradors** [comment]
Thanks for comments @garrett and @jelly !

With everything we have discussed...here the result:

![Peek 2022-07-03 19-36](https://user-images.githubusercontent.com/6016929/177051201-1018a6f2-fac6-4bd4-a2f2-9b725ed64a32.gif)

I would put some tooltips/texthelpers/palceholders...for example in wireguard endpoint, network manager waits for <IP: HOST> format. How do you see it? 

If looks fine, I continue with openvpn.... and maybe [fortinetSSL](https://wiki.gnome.org/Projects/NetworkManager/VPN)? It doesn't comes by default like WireGuard, but it's an installable plugin like openvpn.
I have in mind to check if vpn-provider is available when the modal is loading. If not present, will show a button to install it like others sites from cockpit.

Have nice init(week)!

**Comment by jelly** [comment]
Looks good, I think it makes sense to to first make a PR for just the wireguard functionality and then in future PR's add other VPN solutions.

I do wonder if we don't need a separate section for your existing VPN profiles (from networkmanager). As how else does one activate/de-activate and edit existing profiles?

**Comment by gil-obradors** [comment]
Like this?

![Peek 2022-07-04 19-18](https://user-images.githubusercontent.com/6016929/177197516-d885efa9-437d-4d6f-8132-043f342ca52c.gif)

For enable/disable/edit this user-interface is 100% reusable courtesy of NM objects :smile: 

![imatge](https://user-images.githubusercontent.com/6016929/177197711-2c4be8ee-d6c7-4401-bb97-8d4dea2699b1.png)

**Comment by gil-obradors** [comment]
OK :)

I notice that through install_dialog and `manifest.json` We can install the package on Ubuntu,Debian,Fedora,Arch...
https://github.com/cockpit-project/cockpit/blob/3f5d4c87ef6090cb3174c402a6b64ba9e0c6723d/pkg/lib/cockpit-components-install-dialog.jsx#L55

But How We deal with distributions that need to activate other repositories (RHEL,Centos...)?
https://www.wireguard.com/install/

**Comment by gil-obradors** [comment]
Can be an acceptable work around focus on Ubuntu / Debian / Fedora for self-install  behavior on ADD Wireguard.
And for distributions that do not serve wireguard on main stream, show a message asking for manual install?

**Comment by jelly** [comment]
> OK :)
> 
> I notice that through install_dialog and `manifest.json` We can install the package on Ubuntu,Debian,Fedora,Arch...
> 
> https://github.com/cockpit-project/cockpit/blob/3f5d4c87ef6090cb3174c402a6b64ba9e0c6723d/pkg/lib/cockpit-components-install-dialog.jsx#L55
> 
> But How We deal with distributions that need to activate other repositories (RHEL,Centos...)? https://www.wireguard.com/install/

I personally don't want to recommend adding an EPEL repo as official instructions, I think we should only enable it on modern systems which have it. What we require for wireguad in cockpit should be the following:

* not sure what your code does but if it calls `wg genkey` then we need `wireguard-tools` and we can use the `install_dialog` with manifest for it.
* For networkmanager we require [version 1.16](https://blogs.gnome.org/thaller/2019/03/15/wireguard-in-networkmanager/). It seems that most of the distros we support have this version.
* Kernel could in theory not be build with wireguard support.

> Can be an acceptable work around focus on Ubuntu / Debian / Fedora for self-install behavior on ADD Wireguard. And for distributions that do not serve wireguard on main stream, show a message asking for manual install?

Yes, we don't show anything for things we don't support, it's acceptable to require a modern kernel for this feature and we need NetworkManager > 1.16.

So in short, I guess we want to install wireguard-tools? And then allow users to configure wireguard?

**Comment by gil-obradors** [comment]
Hi!  Working on it.

This is the current scenario:

- New section with Interface Vpn if any detected.
- Add wireguard button conditional to having kernel >= 5.6
- Generate Keys button conditional to have installed wireguard-tools. Distributions without this package will not  be able to auto-generate. Need to manual typing.

Questions:
- Can a user-admin see private key configured in connection? 
Yes, for me. In NM you can. But I have not found how to get this key-value because is protected.
```
➜  ~ nmcli connection show wg1 | grep wireguard 
connection.type:                        wireguard
wireguard.private-key:                  <hidden>
wireguard.private-key-flags:            0 (cap)
wireguard.listen-port:                  0
wireguard.fwmark:                       0x0
wireguard.peer-routes:                  sí
wireguard.mtu:                          0
wireguard.ip4-auto-default-route:       -1 (default)
wireguard.ip6-auto-default-route:       -1 (default)
```
This setting is hiiden by default. But can be shown with -show-secrets :
```
➜   nmcli -show-secrets connection show wg1 | grep wireguard
connection.type:                        wireguard
wireguard.private-key:                  ULQiG2CUuR4/BpYlXj081Y12xYApQ/1b0fBEsUosl30=
wireguard.private-key-flags:            0 (cap)
wireguard.listen-port:                  0
wireguard.fwmark:                       0x0
wireguard.peer-routes:                  sí
wireguard.mtu:                          0
wireguard.ip4-auto-default-route:       -1 (default)
wireguard.ip6-auto-default-route:       -1 (default)
```
I have inspected the function that loads network-manager dbus tree to frontend but wireguard object like `connection`,`ipv4`,`ipv6` isn't present ( and then wireguard.private-key not too)
https://github.com/cockpit-project/cockpit/blob/b971699bfb1311db6420d1b2aefcea1014c7531c/pkg/networkmanager/interfaces.js#L480


- Can an user-admin see public key configured in connection, after being configured? 

Public key value is not stored and not showed in NetworkManager.
I see that `wg` tool wich comes with `wireguard-tools` give you public key of an interface configured. Need to investigate more how to retrive it... maybe can be found inside `/proc` as general for all distributions...  any idea?


- Endpoints must be configured like IP:PORT, and allowed addresses must be a net comma separated. 

How does the project deal with this fields that a newbie-admin may be fall configuring them. Tooltips,comments, or nothing because Network Manger will bring the error message?

![Peek 2022-07-21 18-26](https://user-images.githubusercontent.com/6016929/180265998-f26bd99f-f2f3-4b7a-b14f-ab44c68672dd.gif)

**Comment by jelly** [comment]
> * New section with Interface Vpn if any detected.
> * Add wireguard button conditional to having kernel >= 5.6
> * Generate Keys button conditional to have installed wireguard-tools. Distributions without this package will not  be able to auto-generate. Need to manual typing.
> 
> Questions:
> 
> * Can a user-admin see private key configured in connection?
>   Yes, for me. In NM you can. But I have not found how to get this key-value because is protected.
> 
> ```
> ➜  ~ nmcli connection show wg1 | grep wireguard 
> connection.type:                        wireguard
> wireguard.private-key:                  <hidden>
> wireguard.private-key-flags:            0 (cap)
> wireguard.listen-port:                  0
> wireguard.fwmark:                       0x0
> wireguard.peer-routes:                  sí
> wireguard.mtu:                          0
> wireguard.ip4-auto-default-route:       -1 (default)
> wireguard.ip6-auto-default-route:       -1 (default)
> ```
> 
> This setting is hiiden by default. But can be shown with -show-secrets :
> 
> ```
> ➜   nmcli -show-secrets connection show wg1 | grep wireguard
> connection.type:                        wireguard
> wireguard.private-key:                  ULQiG2CUuR4/BpYlXj081Y12xYApQ/1b0fBEsUosl30=
> wireguard.private-key-flags:            0 (cap)
> wireguard.listen-port:                  0
> wireguard.fwmark:                       0x0
> wireguard.peer-routes:                  sí
> wireguard.mtu:                          0
> wireguard.ip4-auto-default-route:       -1 (default)
> wireguard.ip6-auto-default-route:       -1 (default)
> ```
> 
> I have inspected the function that loads network-manager dbus tree to frontend but wireguard object like `connection`,`ipv4`,`ipv6` isn't present ( and then wireguard.private-key not too)
> 
> https://github.com/cockpit-project/cockpit/blob/b971699bfb1311db6420d1b2aefcea1014c7531c/pkg/networkmanager/interfaces.js#L480

Ok, that's a bummer, I also see no wireguard support in GNOME itself but I can import a connection file. This blog post says the DBus API should support it hmmm.

I did find:

https://developer-old.gnome.org/NetworkManager/stable/gdbus-org.freedesktop.NetworkManager.Device.WireGuard.html

And this can be exposed in cockpit with the following patch:

```
--- a/pkg/networkmanager/interfaces.js
+++ b/pkg/networkmanager/interfaces.js
@@ -972,7 +972,8 @@ export function NetworkManagerModel() {
             "org.freedesktop.NetworkManager.Device.Bond",
             "org.freedesktop.NetworkManager.Device.Team",
             "org.freedesktop.NetworkManager.Device.Bridge",
-            "org.freedesktop.NetworkManager.Device.Vlan"
+            "org.freedesktop.NetworkManager.Device.Vlan",
+            "org.freedesktop.NetworkManager.Device.WireGuard"
         ],

         props: {
@@ -992,6 +993,10 @@ export function NetworkManagerModel() {
             Carrier:              { def: true },
             Speed:                { },
             Managed:              { def: false },
+            // WireGuard
+            PublicKey:            { },
+            ListenPort:           { },
+            FwMark:               { },
             // See below for "Members"
         },
```

But that doesn't give the information you want or should likely use.

> * Can an user-admin see public key configured in connection, after being configured?
> 
> Public key value is not stored and not showed in NetworkManager. I see that `wg` tool wich comes with `wireguard-tools` give you public key of an interface configured. Need to investigate more how to retrive it... maybe can be found inside `/proc` as general for all distributions... any idea?

As shown above it can be retrieved.

> 
> * Endpoints must be configured like IP:PORT, and allowed addresses must be a net comma separated.
> 
> How does the project deal with this fields that a newbie-admin may be fall configuring them. Tooltips,comments, or nothing because Network Manger will bring the error message?

Patternfly has tooltips and validation so this is something we can add and do in other places in the UI

> 
> ![Peek 2022-07-21 18-26](https://user-images.githubusercontent.com/6016929/180265998-f26bd99f-f2f3-4b7a-b14f-ab44c68672dd.gif) [ ![Peek 2022-07-21 18-26](https://user-images.githubusercontent.com/6016929/180265998-f26bd99f-f2f3-4b7a-b14f-ab44c68672dd.gif) ](https://user-images.githubusercontent.com/6016929/180265998-f26bd99f-f2f3-4b7a-b14f-ab44c68672dd.gif) [ ](https://user-images.githubusercontent.com/6016929/180265998-f26bd99f-f2f3-4b7a-b14f-ab44c68672dd.gif)

**Comment by jelly** [comment]
I think the dialog should use the connection settings API
```
[jelle@t14s][~/projects/cockpit/main]%busctl call org.freedesktop.NetworkManager /org/freedesktop/NetworkManager/Settings/36 org.freedesktop.NetworkManager.Settings.Connection GetSettings
a{sa{sv}} 5 "connection" 6 "id" s "home" "interface-name" s "home" "permissions" as 0 "timestamp" t 1660570370 "type" s "wireguard" "uuid" s "$UUID" "wireguard" 1 "peers" aa{sv} 1 4 "public-key" s "PUBKEY" "endpoint" s "8.8.8.8:10" "persistent-keepalive" u 25 "allowed-ips" as 1 "10.0.0.0/24" "ipv4" 6 "address-data" aa{sv} 1 2 "address" s "10.200.200.4" "prefix" u 32 "addresses" aau 1 3 339339 32 0 "dns-search" as 0 "method" s "manual" "route-data" aa{sv} 0 "routes" aau 0 "ipv6" 6 "address-data" aa{sv} 0 "addresses" a(ayuay) 0 "dns-search" as 0 "method" s "disabled" "route-data" aa{sv} 0 "routes" a(ayuayu) 0 "proxy" 0
```

**Comment by garrett** [comment]
> I also see no wireguard support in GNOME itself

It's supposedly a feature in GNOME 43, which has just hit beta.

https://9to5linux.com/gnome-43-alpha-released-to-kick-off-guadec-2022-in-guadalajara-mexico

> The upcoming GNOME desktop release also comes with a new Sidebar and the ability to zoom the Week view for GNOME Calendar, support for composite emoji in GNOME Characters, **improved handling of VPN connections (especially for WireGuard)**, more Wayland improvements, and more GTK4 and libadwaita ports.

However, I'm on the latest GNOME OS nightly in GNOME Boxes (Nightly) and I don't see it:

![image](https://user-images.githubusercontent.com/10246/184831177-d3b90c9b-4f37-4395-9d2a-59ac95e4cb81.png)

Perhaps it requires something in the stack that isn't in GNOME OS (yet)?

I did try Fedora Silverblue rebased to Rawhide and did install all the wireguard related packages Fedora ships with (`wireguard-tools` `wgctrl`), just in case. But it doesn't show wireguard either. (It does have an SSH connection as VPN which seems nifty though and PPTP is there too.)

![Screenshot from 2022-08-16 10-30-00](https://user-images.githubusercontent.com/10246/184834825-42f8e1d7-9d7f-49cf-8e4c-5d8251c327d7.png)

---

I did a little digging around and found the original PR for GNOME Settings about wireguard creating and editing support — with screenshots — @ https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/1125, but it was closed in favor of https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/1364 which is still open (and marked as WIP). 

Here are the screenshots, for comparison (you can tell it's from an older PR as these aren't using the newer libadwaita style):

![image](https://user-images.githubusercontent.com/10246/184836983-d4be128d-f31a-496d-aae9-91bc2270520c.png)

![image](https://user-images.githubusercontent.com/10246/184837020-d24ecbd4-efa7-429e-b583-793fb3a2632a.png)

![image](https://user-images.githubusercontent.com/10246/184837074-af94648b-f78a-4193-8581-b75a381e4afa.png)

![image](https://user-images.githubusercontent.com/10246/184837114-af00634c-9c52-4901-b0d4-17da00193482.png)


Meanwhile, GNOME Shell did get support for wireguard toggling in VPNs @ https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/1995.

I guess this is what the blog post was talking about? And I guess creating and editing will (eventually) show up in GNOME 44?

**Comment by jelly** [comment]
Oh and FYI, how I found settings dbus calls was by running `sudo busctl monitor > log` and in a different terminal running `nmcli con show home` and you can see how they create a new profile with `nmcli con import type wireguard file /etc/wireguard/home.conf`.

**Comment by jelly** [comment]
> > I also see no wireguard support in GNOME itself
> 
> It's supposedly a feature in GNOME 43, which has just hit beta.
> 
> https://9to5linux.com/gnome-43-alpha-released-to-kick-off-guadec-2022-in-guadalajara-mexico
> 
> > The upcoming GNOME desktop release also comes with a new Sidebar and the ability to zoom the Week view for GNOME Calendar, support for composite emoji in GNOME Characters, **improved handling of VPN connections (especially for WireGuard)**, more Wayland improvements, and more GTK4 and libadwaita ports.
> 
> However, I'm on the latest GNOME OS nightly in GNOME Boxes (Nightly) and I don't see it:
> 
> ![image](https://user-images.githubusercontent.com/10246/184831177-d3b90c9b-4f37-4395-9d2a-59ac95e4cb81.png)
> 
> Perhaps it requires something in the stack that isn't in GNOME OS (yet)?
> 
> I did try Fedora Silverblue rebased to Rawhide and did install all the wireguard related packages Fedora ships with (`wireguard-tools` `wgctrl`), just in case. But it doesn't show wireguard either. (It does have an SSH connection as VPN which seems nifty though and PPTP is there too.)
> 
> ![Screenshot from 2022-08-16 10-30-00](https://user-images.githubusercontent.com/10246/184834825-42f8e1d7-9d7f-49cf-8e4c-5d8251c327d7.png)
> 
> I did a little digging around and found the original PR for GNOME Settings about wireguard creating and editing support — with screenshots — @ https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/1125, but it was closed in favor of https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/1364 which is still open (and marked as WIP).
> 
> Here are the screenshots, for comparison (you can tell it's from an older PR as these aren't using the newer libadwaita style):
> 
> ![image](https://user-images.githubusercontent.com/10246/184836983-d4be128d-f31a-496d-aae9-91bc2270520c.png)
> 
> ![image](https://user-images.githubusercontent.com/10246/184837020-d24ecbd4-efa7-429e-b583-793fb3a2632a.png)
> 
> ![image](https://user-images.githubusercontent.com/10246/184837074-af94648b-f78a-4193-8581-b75a381e4afa.png)
> 
> ![image](https://user-images.githubusercontent.com/10246/184837114-af00634c-9c52-4901-b0d4-17da00193482.png)
> 
> Meanwhile, GNOME Shell did get support for wireguard toggling in VPNs @ https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/1995.
> 
> I guess this is what the blog post was talking about? And I guess creating and editing will (eventually) show up in GNOME 44?

I assume they first added Wireguard support in NetworkManager and importing of profiles. And then gradually support it in GNOME itself \o/

Already being able to toggle it in GNOME Shell is a big win for me :-)

**Comment by gil-obradors** [comment]
happy to read you!
Summing up:

- (Wireguard) _network-vpn-interface-modal_ to NM is :ok: 
- NM to  (wireguard) _network-vpn-interface-modal_ have problems:

1. Need to read object Device.Wireguard.  We want `ListenPort` and `PublicKey`
I understand with that patch We are able to load proprieties object `Device.Wireguard`?
```
--- a/pkg/networkmanager/interfaces.js
+++ b/pkg/networkmanager/interfaces.js
@@ -972,7 +972,8 @@ export function NetworkManagerModel() {
             "org.freedesktop.NetworkManager.Device.Bond",
             "org.freedesktop.NetworkManager.Device.Team",
             "org.freedesktop.NetworkManager.Device.Bridge",
-            "org.freedesktop.NetworkManager.Device.Vlan"
+            "org.freedesktop.NetworkManager.Device.Vlan",
+            "org.freedesktop.NetworkManager.Device.WireGuard"
         ],

         props: {
@@ -992,6 +993,10 @@ export function NetworkManagerModel() {
             Carrier:              { def: true },
             Speed:                { },
             Managed:              { def: false },
+            // WireGuard
+            PublicKey:            { },
+            ListenPort:           { },
+            FwMark:               { },
             // See below for "Members"
         },
```
I can't see them...

3. Need to call GetSecrets for obtain private-key from `Wireguard-Instance `
I have found that there's a GetSecrets Method for that, and privatekey is stored there.
How Can I call this method from cockpit?
```
➜  ~ dbus-send --system --print-reply --dest=org.freedesktop.NetworkManager /org/freedesktop/NetworkManager/Settings/13 org.freedesktop.NetworkManager.Settings.Connection.GetSecrets string:"wireguard"
method return time=1661020841.177166 sender=:1.12 -> destination=:1.19778 serial=84779 reply_serial=2
   array [
      dict entry(
         string "connection"
         array [
         ]
      )
      dict entry(
         string "wireguard"
         array [
            dict entry(
               string "peers"
               variant                   array [
                     array [
                        dict entry(
                           string "public-key"
                           variant                               string "xq/101tmyY6nM9IgXobQ6MZr5+YK+0gLkFUKjE72cHo="
                        )
                     ]
                  ]
            )
            dict entry(
               string "private-key"
               variant                   string "0IiqhN9DBmKsXx5mjknRbSgu8wLybzqUsV1+1RwsLnI="
            )
         ]
      )
      dict entry(
         string "ipv4"
         array [
         ]
      )
      dict entry(
         string "ipv6"
         array [
         ]
      )
      dict entry(
         string "proxy"
         array [
         ]
      )
   ]
```
Or as a workaround, We can call from CLI with cockpit.console...
```
➜  ~ nmcli --show-secrets -f wireguard.private-key connection show wireguard
wireguard.private-key:                  0IiqhN9DBmKsXx5mjknRbSgu8wLybzqUsV1+1RwsLnI=
➜  ~ 
```

**Comment by jelly** [comment]
With:

```
diff --git a/pkg/networkmanager/interfaces.js b/pkg/networkmanager/interfaces.js
index 277ede9de..4795b681f 100644
--- a/pkg/networkmanager/interfaces.js
+++ b/pkg/networkmanager/interfaces.js
@@ -972,7 +972,8 @@ export function NetworkManagerModel() {
             "org.freedesktop.NetworkManager.Device.Bond",
             "org.freedesktop.NetworkManager.Device.Team",
             "org.freedesktop.NetworkManager.Device.Bridge",
-            "org.freedesktop.NetworkManager.Device.Vlan"
+            "org.freedesktop.NetworkManager.Device.Vlan",
+            "org.freedesktop.NetworkManager.Device.WireGuard"
         ],

         props: {
@@ -992,6 +993,8 @@ export function NetworkManagerModel() {
             Carrier:              { def: true },
             Speed:                { },
             Managed:              { def: false },
+            PublicKey:            { },
+            ListenPort:           { },
             // See below for "Members"
         },

diff --git a/pkg/networkmanager/network-main.jsx b/pkg/networkmanager/network-main.jsx
index 09402b8be..2949ded6b 100644
--- a/pkg/networkmanager/network-main.jsx
+++ b/pkg/networkmanager/network-main.jsx
@@ -71,6 +71,7 @@ export const NetworkPage = ({ privileged, operationInProgress, usage_monitor, pl
             return;

         const dev = iface.Device;
+        console.log(iface);
         const show_traffic = (dev && (dev.State == 100 || dev.State == 10) && dev.Carrier === true);

         plot_ifaces.push(iface.Name);
```

I do see the ListenPort/PublicKey but **only** when the device is active, not inactive. So that's not super useful. Probably better  to get it from the settings.

**Comment by jelly** [comment]
> 3. Need to call GetSecrets for obtain private-key from `Wireguard-Instance `
>    I have found that there's a GetSecrets Method for that, and privatekey is stored there.
>    How Can I call this method from cockpit?

This is possible, see for example:

https://github.com/cockpit-project/cockpit/blob/main/pkg/networkmanager/interfaces.js#L774

`objpath` resolves to `path: "/org/freedesktop/NetworkManager/Settings/29"` for me, so you need to call

```
client.call("/org/freedesktop/NetworkManager/Settings/29", "org.freedesktop.NetworkManager.Settings.Connection", "GetSecrets")
```

Haven't tested it myself.

**Comment by gil-obradors** [comment]
Thanks Jelly!
With all this information I can already see an initial feasible version.
I will try to put it in the oven this week!

**Comment by jelly** [comment]
Cool! Feel free to make a draft PR then I can take a look at the code.

**Comment by gil-obradors** [comment]
I stopped my work when I ran into trouble trying to recover credentials from Wireguard via d-bus. May be remote commands from shell can be a workaround, but not via d-bus interfaces ( at least when I managed to get ).

Some work from screnshoots is [here ](https://github.com/gil-obradors/cockpit)

**Comment by martinpitt** [comment]
@subhoghoshX is working on this topic as Google Summer of Code project. See PR #19024 for a first draft.

**Comment by mispp** [comment]
So, this works well. Would be nice to have two things:
- keepalive option in the gui
- list of textboxes for allowed ips instead of one textbox

**Comment by subhoghoshX** [comment]
Thanks for the feedback @mispp. Yes having a PersistentKeepalive is a nice to have in a few situations. Created an issue #19491.

I'm not clear about the benefit of having separate text inputs for each allowed-ips. Is having a comma/space separated list of IPs confusing? or is the problem that the width of the input field is too small? In the later case it can be moved to a separate row if PersistentKeepalive is added. We can't fit four fields in a row anyway.

**Comment by mispp** [comment]
Benefits I see:
- better overview of what is allowed when multiple subnets are entered
- easier entry / validation

This is definitely not a must. If this is not implemented, can you at least make an example below the textbox? My first try included spaces after comma which got invalidated.

Making texbox wider would help.

Thanks for making an issue for keepalive.

**Comment by mispp** [comment]
One more thing which would be nice and it would make many tools obsolete: adding a download/send configuration for the opposite peer. Think in terms of server (machine running this cockpit) and other peers (road warriors). Configuration for road warriors could be downloaded here in ini Format what Android App wants

**Comment by Venefilyn** [comment]
Closing as complete as of #19024

Any other features relating to the WireGuard implementation or bugs feel free to open a new issue
